### PR TITLE
PLF-8323: Favorites and Folksonomy default folders should be created …

### PR DESCRIPTION
**Pull request by fmaatoug**
Favorites and folksonomy folders should be created Or else the UI rendering will be broken and the cases of their paths not found in configuration should be handled gracefully

### CODE REVIEW

File: core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java [hunk 1]
```diff
@@ -183,7 +183,15 @@ private Node getUserFavoriteFolder(String userName) throws Exception {
...
-    return userNode.getNode(favoritePath);
+    if (favoritePath == null) {
+      return null;
+    }
+    try {
+      return userNode.getNode(favoritePath);
+    } catch (PathNotFoundException e) {
+      Node userFavoriteNode = createFavoriteFolder(userName);
+      return userFavoriteNode;
+    }
```
Comment (thomasdelhomenie) [inline core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: The logic should not be managed with exception IMO, we should instead use `if(!userNode.hasNode(favoritePath)) { ... }` to check if the folder is created.

File: core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java [hunk 2]
```diff
@@ -183,7 +183,15 @@ private Node getUserFavoriteFolder(String userName) throws Exception {
...
       nodeHierarchyCreator.getUserNode(sessionProviderService.getSystemSessionProvider(null), userName);
     String favoritePath = nodeHierarchyCreator.getJcrPath(FAVORITE_ALIAS);
-    return userNode.getNode(favoritePath);
+    if (favoritePath == null) {
+      return null;
+    }
+    try {
+      return userNode.getNode(favoritePath);
+    } catch (PathNotFoundException e) {
+      Node userFavoriteNode = createFavoriteFolder(userName);
```
Comment (thomasdelhomenie) [inline core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: If no folder exists, it means there is no favorites, so we could return an empty collection without creating the folder node. It avoids to create an useless folder in the case the user never use favorites. WDYT ?

File: core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java [hunk 3]
```diff
@@ -781,7 +792,39 @@ private Node getUserFolksonomyFolder(String userName) throws Exception {
...
-    return userNode.getNode(folksonomyPath);
+    if (folksonomyPath == null) {
+      return null;
+    }
+    try {
+      return userNode.getNode(folksonomyPath);
+    } catch (PathNotFoundException e) {
+      Node userFolksonomyNode = createFolksonomyFolder(userName);
+      return userFolksonomyNode;
+    }
```
Comment (thomasdelhomenie) [inline core/services/src/main/java/org/exoplatform/services/cms/folksonomy/impl/NewFolksonomyServiceImpl.java]: Same remarks than for favorites.

File: core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java (hunk 2 above)
Comment (fmaatoug) [reply core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: I agree i think it's useless to recreate it if the user already removed it

File: core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java (hunk 1 above)
Comment (fmaatoug) [reply core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: Agreed

File: packaging/wcm/webapp/src/main/webapp/WEB-INF/conf/dms-extension/dms/artifacts/scripts/ecm-explorer/action/AddToFavorites.groovy [hunk 4]
```diff
@@ -65,6 +65,10 @@ public class AddToFavoriteScript implements CmsScript {
       String userID = ConversationState.getCurrent().getIdentity().getUserId();
       if(userID.equals(IdentityConstants.ANONIM) || userID.equals(IdentityConstants.SYSTEM) || (activateUser != null && activateUser.equals(IdentityConstants.SYSTEM))) return;
       ExtendedNode userNode = (ExtendedNode)nodeHierarchyCreator_.getUserNode(WCMCoreUtils.getSystemSessionProvider(), userID);
+      if (nodeHierarchyCreator_.getJcrPath(FAVORITE_ALIAS) == null){
+        LOG.debug("Favorite path is not configured. The feature 'Favorite Documents' will be ignored")
+        return;
+      }
       String favoritePath = nodeHierarchyCreator_.getJcrPath(FAVORITE_ALIAS);
```
Comment (thomasdelhomenie) [inline packaging/wcm/webapp/src/main/webapp/WEB-INF/conf/dms-extension/dms/artifacts/scripts/ecm-explorer/action/AddToFavorites.groovy]: This line should be moved before the if, so we can check if favoritePath is null and therefore avoid calling `nodeHierarchyCreator_.getJcrPath(FAVORITE_ALIAS)` twice.

File: core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java [hunk 5]
```diff
@@ -183,6 +183,9 @@ private Node getUserFavoriteFolder(String userName) throws Exception {
     Node userNode =
       nodeHierarchyCreator.getUserNode(sessionProviderService.getSystemSessionProvider(null), userName);
     String favoritePath = nodeHierarchyCreator.getJcrPath(FAVORITE_ALIAS);
+    if (favoritePath == null || !userNode.hasNode(favoritePath)) {
+      return null;
+    }
```
Comment (thomasdelhomenie) [inline core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: Can we add a unit test to check that the method does not raise an exception when the folder does not exist ?

Comment (fmaatoug) [reply core/services/src/main/java/org/exoplatform/services/cms/documents/impl/FavoriteServiceImpl.java]: yes i think we should test this behaviour

**Pull request merged**